docs/tutorial.md

```diff
@@ -0,0 +1,190 @@
+#### stage 3 Prepare Wenet data format
+
+This stage generates a single wenet format file including all the input/output information needed by neural network training/evaluation.
+
+See the generated training feature file in `fbank_pitch/train/format.data`.
+
+
+In the wenet format file , each line record a data sample of seven tab splited columns. For example, a line is like this(repalce tab with newline here):
+
+```
+utt:BAC009S0764W0121
+feat:/l/wenet/examples/aishell/s0/fbank_pitch/test/data/raw_fbank_pitch_test.1.ark:17
+feat_shape:418,83
+text:甚至出现交易几乎停滞的情况
+token:甚 至 出 现 交 易 几 乎 停 滞 的 情 况
+tokenid:2474 3116 331 2408 82 1684 321 47 235 2199 2553 1319 307
+token_shape:13,4233
+```
+
+In the future, wenet will also support using raw wavs as input and use torchaudio to extract the feature just-in-time in dataloader.
+
+
+#### stage 4 Neural Network training
+
+The NN model is training in this step.
+
+- multi-GPU mode
+
+If using DDP mode for multi-GPU, we suggest using `dist_backend="nccl"`. If the NCCL does not work, try using `gloo` or use `torch==1.6.0`
+Set the GPU ids in CUDA_VISIBLE_DEVICES. For example, set `export CUDA_VISIBLE_DEVICES="0,1,2,3,6,7"` to use card 0,1,2,3,6,7.
+
+- BreakPoint training
+
+If your experiment is terminated after running several epochs for some reasons(e.g. the GPU is accidentally used by other people and is out-of-memory ), you could continue the training on a checkpoint model. Just check the finished epoch in `exp/your_exp/` and set  `checkpoint=exp/your_exp/$n.pt` and run the `run.sh --stage 4`. Then the training will continue from the $n+1.pt
+
+- Config
+
+The config of neural network structure, optimization parameter, loss parameters, and dataset could be set in a YAML format file.
+
+In `conf/`,  we provide several models like transformer and conformer. see `conf/train_conformer.yaml` for reference.
+
+- Use Tensorboard
+
+The training takes several hours. The actual hours depend on your GPU number and type. In an 8 card 2080 Ti machine, it will take about less than one day for 50 epochs.
+You could use the tensorboard to monitor the loss.
+
+```
+tensorboard --logdir tensorboard/$your_exp_name/ --port 12598 --bind_all
+```
+
+#### stage 5 Recognize wav using the trained model
+
+This stage shows how to recognize a set of wavs into text. It also shows how to do the model average.
+
+- Average model
+
+if `${average_checkpoint}` is `true`, the best `${average_num}` model on cross validation set will be averaged to generated a boosted model and used for recognition.
+
+- Decoding
+
+Recognition is also called decoding or inference. The function of the NN will be applied on the input acoustic feature sequence to out a text sequence.
+
+Four decoding methods are provided in wenet:
+
+* ctc_greedy_search: encoder + CTC greedy search
+* ctc_prefix_beam_search:  encoder + CTC prefix beam search
+* attention: encoder + attention-based decoder decoding
+* attention_rescoring: rescoring on the ctc candidates
```

Comment:
rescoring the ctc candidates from ctc prefix beam search with encoder output on attention-based decoder.